automatic/nexus-repository/tools/helpers.ps1

```diff
@@ -105,3 +97,189 @@ function Wait-NexusAvailability {
+}
+
+function Test-NexusDatabaseType {
+  <#
+    .Synopsis
+      Checks the currently configured database type in use by Nexus
+
+    .Example
+      Test-NexusDatabaseType -is "OrientDb"
+  #>
+  [OutputType([bool])]
+  [CmdletBinding()]
+  param(
+    [ValidateSet("OrientDb","H2","PostgreSQL")]
+    [Parameter(Mandatory)]
+    [Alias('is','eq')]
+    [string]$Type,
+    [string]$ProgramDir = (Join-Path $env:ProgramData 'nexus'),
+    [string]$DataDir = (Join-Path $env:ProgramData 'sonatype-work')
+  )
+  $JdbcUrl = if ($env:NEXUS_DATASTORE_NEXUS_JDBCURL) {
+    $env:NEXUS_DATASTORE_NEXUS_JDBCURL
+  } elseif ((Test-Path "$DataDir\etc\fabric\nexus-store.properties") -and ($Properties = Select-String -Path "$DataDir\etc\fabric\nexus-store.properties" -Pattern '^jdbcUrl=(?<Url>.+)$')) {
+    $Properties.Matches.Groups[1].Value
+  } elseif ($Properties = Select-String -Path "$ProgramDir\bin\nexus.vmoptions" -Pattern '^(?!#)\W*-Dnexus\.datastore\.nexus\.jdbcUrl=(?<Url>.+)') {
+    $Properties.Matches.Groups[1].Value
+  }
+
+  $FoundType = if ($JdbcUrl) {
+    if ($JdbcUrl -match 'jdbc\\?:h2') {
+      "H2"
+    } elseif ($JdbcUrl -match 'jdbc\\?:postgresql') {
+      "PostgreSQL"
+    }
+  } else {
+    # No external database found - probably...
+    "OrientDb"
+  }
+
+  $Type -eq $FoundType
+}
+
+function New-NexusOrientDbBackup {
+  <#
+    .Synopsis
+      Produces a nearly-authentic backup file for all specified database types
+
+    .Example
+      New-NexusOrientDbBackup
+      # Backs up all the databases needed for a migration.
+
+    .Example
+      New-NexusOrientDbBackup -IncludedDatabases 'config' -DestinationPath $PWD
+      # Backs up the config database to the present working directory.
+  #>
+  [CmdletBinding(SupportsShouldProcess)]
+  param(
+    [string]$DestinationPath = (Join-Path $env:TEMP 'sonatype-backup'),
+    [string]$DataDir = (Join-Path $env:ProgramData 'sonatype-work'),
+    [string]$BaseName = "-$(Get-Date -Format 'yyyy-MM-dd-HH-mm-ss')-$(Get-NexusVersion).bak",
+    [string[]]$IncludedDatabases = @(
+      # "analytics"
+      "component"
+      "config"
+      "security"
+    ),
+    [string]$ServiceName = "nexus",
+    [switch]$PassThru
+  )
+  if (($Service = Get-Service $ServiceName -ErrorAction SilentlyContinue).Status -ne 'Stopped') {
+    if ($PSCmdlet.ShouldProcess($ServiceName, "Stopping Service")) {
+      Stop-Service $ServiceName
+    }
+  }
+  try {
+    if (-not (Test-Path $DestinationPath) -and $PSCmdlet.ShouldProcess($DestinationPath, "Ensuring Folder")) {
+      $null = New-Item -Path $DestinationPath -ItemType Directory -Force
+    }
+    $PreviousProgress, $ProgressPreference = $ProgressPreference, "SilentlyContinue"
+    foreach ($Database in $IncludedDatabases) {
+      Push-Location (Join-Path $DataDir "nexus3/db/$Database")
+      $ArchiveArgs = @{
+        Path = Get-ChildItem -Path (Join-Path $DataDir "nexus3/db/$Database/*") -Exclude "cache.stt" -File
```

Comment:
My understanding is that to ensure the database in a consistent state, the Nexus task to export the database should be used. I'm just concerned that this could be creating a backup with the database not working correctly.

Reply:
That's why I'm shutting down the service.

To be clear, I have both:
- Used BeyondCompare to check that my backups are file-identical (the only difference is the compression algo used)
- Tested a _bunch_ of times that migrations succeed, and import all of the data (that I can see) successfully

Reply:
Resolving this off, as I hope I answered the question / addressed the concern. If not, please let me know.